### PR TITLE
INC-1321: Automate KPIs

### PR DESCRIPTION
**Pull request by xoen**
Monthly task to calculate/save:
1. the number of reviews occurred in the previous, complete, month
2. the number of prisoners reviewed in the previous, complete, month
3. the number of prisoners overdue a review at the point when the task runs

First 2 can be calculate with a single query that look at the reviews in the previous month.
3rd is a bit more complicated because it gets the data from the DB but then it filter out people no longer in prison using Prison API/Offender Search API.

The automated task runs on the 1st of the month at 3:00am in production.
It also run on the 1st working day of the month at 10:00am in preprod/prod (because Offender Search API is switched off in dev/preprod during non-working hours).

Shedlock is used to prevent the task from running multiple times in all the application instances.

### CODE REVIEW

File: src/main/resources/application.yml [hunk 1]
```diff
@@ -2,6 +2,11 @@ info.app:
   name: Hmpps Incentives API
   version: 1.0
 
+task:
+  update-kpis:
+    # Every 1st of the month at 3:00am
```
Comment (ushkarev) [inline src/main/resources/application.yml:7]: oh actually sounds v sensible to do the "how many overdue reviews?" check at night cos nobody is part-way through whatever they've planned for the day.

Comment (xoen) [reply src/main/resources/application.yml:7]: Yeah, and it's a snapshot anyway so as long as it's a number taken around the same time it doesn't really matter the precise time when it's taken.

BTW: I had a chat with Steve this morning and he concurs. He's happy with this and he agrees it doesn't have to be taken separately on the last day of the month.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/dto/prisonapi/PrisonAPI.kt [hunk 2]
```diff
@@ -5,6 +5,14 @@ import uk.gov.justice.digital.hmpps.incentivesapi.service.PrisonerInfoForNextRev
 import java.time.LocalDate
 import java.time.LocalDateTime
 
+data class Prison(
+  val agencyId: String,
+  val description: String,
+  val longDescription: String,
```
Comment (ushkarev) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/dto/prisonapi/PrisonAPI.kt:11]: potential danger: this field in nullable (at least according to swagger docs; i didn't check source)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonApiService.kt [hunk 3]
```diff
@@ -73,4 +74,12 @@ class PrisonApiService(
       .retrieve()
       .awaitBody()
   }
+
+  suspend fun getActivePrisons(useClientCredentials: Boolean = false): List<Prison> {
```
Comment (ushkarev) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonApiService.kt:78]: `useClientCredentials` is a bit of a mystery to me, but i know there's an effort to use "system" tokens and their roles. just leaving this note here to maybe explore using the `hmpps-incentives-api` token and giving it any extra roles if necessary

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/Kpi.kt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+data class Kpi(
+  @Id
+  val day: LocalDate,
+
+  val overdueReviews: Int,
+  val previousMonthReviewsConducted: Int,
+  val previousMonthPrisonersReviewed: Int,
+
+  @ReadOnlyProperty
+  val whenCreated: LocalDateTime? = null,
```
Comment (ushkarev) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/Kpi.kt:20]: from recollection, our "reactive" setup means you need to ` = LocalDateTime.now()` otherwise whenCreated and whenUpdated are not set…? or did it work for you?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/repository/KpiRepository.kt [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+   */
+  @Query(
+    //language=postgresql
+    """
+    -- this may include people no longer in prison
+    SELECT DISTINCT ON (prisoner_number) booking_id, prisoner_number
+    FROM prisoner_iep_level
+    JOIN next_review_date USING (booking_id)
+    WHERE next_review_date < CURRENT_DATE
+    ;
```
Comment (ushkarev) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/repository/KpiRepository.kt:53]: with `SELECT DISTINCT ON (…)` you need to have an `ORDER BY` clause so as to pick the latest review rather than _some_ review:

`ORDER BY prisoner_number, review_time DESC NULLS LAST`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/Kpi.kt (hunk 4 above)
Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/Kpi.kt:20]: I've copied the model from the others. So it should be the setup that we made it work.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonApiService.kt (hunk 3 above)
Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonApiService.kt:78]: This method has this boolean for consistency with the other methods in this class. But the `KpiService` will use the system token (by passing `true`)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/dto/prisonapi/PrisonAPI.kt (hunk 2 above)
Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/dto/prisonapi/PrisonAPI.kt:11]: Good spot. Even if we don't care for this field for the `KpiService` it could still cause deserialisation issues if some of the prisons have `null` long descriptions. I'll change 👌

Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/dto/prisonapi/PrisonAPI.kt]: Changed `Prison` dto class to have optional `longDescription`: https://github.com/ministryofjustice/hmpps-incentives-api/pull/503/commits/9fc99f7f11299240b9a2e7f330d7d9532f109b1b

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/Kpi.kt (hunk 4 above)
Comment (ushkarev) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/Kpi.kt:20]: the `IncentiveLevel`, for instance, has ` = LocalDateTime.now()`.

however `NextReviewDate` has ` = null` instead so at instantiation we have to do `whenUpdated = LocalDateTime.now(clock)`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/repository/KpiRepository.kt (hunk 5 above)
Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/repository/KpiRepository.kt:54]: Thanks for the explanation earlier Igor. I'm still not 100% sure the ORDER by would make a difference in our case (given we're only interested in the number of overdue prisoners) but I've added just to be on the safe side.

Also good to have a well-defined order given how `DISTINCT ON` work (e.g. it is a glorified group by).

I've also added `review_time` to the `SELECT`, even tho we don't use that data. But the order of the rows returned seem to change so I've deemed to be saver to add that too.

If the query still doesn't look right let me know: https://github.com/ministryofjustice/hmpps-incentives-api/pull/503/commits/aae18d5caca95a4dbdd3a3ef57e3c17e31945181

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/HmppsIncentivesApi.kt [hunk 6]
```diff
@@ -10,6 +13,8 @@ import org.springframework.boot.runApplication
 const val SYSTEM_USERNAME = "INCENTIVES_API"
 
 @SpringBootApplication
+@EnableScheduling
+@EnableSchedulerLock(defaultLockAtMostFor = defaultLockAtLeastFor)
```
[the remark this answers is not in the document]
Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/HmppsIncentivesApi.kt:17]: I was following along this article here: https://www.baeldung.com/shedlock-spring

But I wouldn't be surprised if these annotation worked wherever.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/KpiService.kt [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    println("Number of prisoners across the estate: ${allPrisoners.size}")
+
+    // Get prisoner numbers overdue, "filter" out people no longer in prison and return the count
+    kpiRepository
+      .getPrisonerNumbersOverdueReview()
+      .count { allPrisoners.contains(it.prisonerNumber) }
+  }
+
+  private fun getPrisons(): List<String> = runBlocking {
```
[the remark this answers is not in the document]
Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/KpiService.kt:41]: Mainly because I wanted to make all the other requests in concurrently.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/repository/KpiRepository.kt (hunk 5 above)
Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/repository/KpiRepository.kt:54]: True. Although the little class may be more open for extension in the future. But I'll probably change.

Comment (ushkarev) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/repository/KpiRepository.kt:54]: honestly, i'm happy with a named object!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/HmppsIncentivesApi.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (ushkarev) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/HmppsIncentivesApi.kt:17]: cool beans! i was just checking!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/KpiService.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (ushkarev) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/KpiService.kt:41]: cool

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/repository/KpiRepository.kt (hunk 5 above)
Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/repository/KpiRepository.kt:58]: Hopefully this commit message make some sense 😅: [: Fix query to get overdue prisoners](https://github.com/ministryofjustice/hmpps-incentives-api/pull/503/commits/4896983ef9babbab0ecf9db0e2e493e5820bebef)

I tweaked the test so that the data had that edge case where a single prisoner number had multiple bookings, some with overdue review but some not overdue.

Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/repository/KpiRepository.kt:54]: I've also simplified the return type of that repository method: 7ab3544fe5f049145caf12158079a9e42d0b5835

**Review by ushkarev — APPROVED** [review]
Worth a try, eh? 🤪

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/Kpi.kt (hunk 4 above)
Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/jpa/Kpi.kt:20]: Just to recap our offline chat. Both fields are automatically set to `now()` correctly because of the DB table `DEFAULT now()` (similar to what happen for the next_review_date table/JPA entity).
If we were to update any of these records we'll have explicitly pass the `whenUpdated` field.

I don't think there is anything to do with this tbh.
